Changed DeepTau modules so that they can use either AOD or RECO input

lwezenbe opened the pull request:
#### The DeepTau modules have been adapted to not only handle AOD input but to also be able to process RECO input.\

The following changes have been made:\
- Input tokens for Candidate and Tau collections now use base classes
- taus and candidates are accessed by dynamic_cast and edm::View on the base class objects
- A boolean "is_online" is added to specify whether the input is AOD or RECO
- Electron and muon collections are initialized as empty arrays if is_online is True
- A structure "tauFunc" containing a range of functions was added to calculate both tau ID quantities as well as other quantities that needed recalculation for RECO input. 
- A namespace "candFunc" was introduced containing functions to calculate quantities that need recalculation for RECO input. 
- A structure lightLepFunc was added to initialize the electron and muon collections depending on the value of the boolean "is_online"
- Functions that use taus or candidates have been turned into template functions where the CandidateType and/or TauType are specified

A few small integration tests have been performed to check if the adapted code works for both RECO and miniAOD inputs:\
- RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py was run with a setup without any changes and the version of the setup in this PR over the same 1000 events of the same dataset. The output of both runs was identical.
- A simple setup was made with hltGetConfiguration for the latest HLT menu where the isolation modules were replaced by the adapted DeepTauId module. This setup ran smoothly and the output looked as expected.

mbluj commented:
Thanks Liam! I hope start review of this development ASAP (this week), I hope that @swozniewski will also help. I have already a few general suggestions:
* Probably it can be useful to add "prediscriminat" as in other tauIDs to not run deepTau on useless taus.
* Maybe it makes sense to allow also usage with offline AOD? It will require differentiation between `is_online` and `is_aod`  (the former will be more exclusive than the latter)? Of course for offline AOD one will need implement interfaces for reco:Muon and reco::GsfElectron (basically to be copied from classic against muon and electron discriminators)
* Generally PFTaus can be also build with pat::PackedCandidates, so maybe one should also consider it in this code (possibly relevant for full tau reco at miniAOD when `is_aod` option is implemented).

swozniewski commented:
Yes, I'd also like to review. I was about to ask whether this is already supposed to be reviewed or still in development, but you have given the answer already. I'll try to do it this week, too.

kandrosov commented:
> * Probably it can be useful to add "prediscriminat" as in other tauIDs to not run deepTau on useless taus.

- I think for online pre-discriminant can depend on a path, so it can be added later after the new paths definitions are established. On the other hand, for offline AOD it makes sense to apply the same preselection as for miniAOD.

> * Maybe it makes sense to allow also usage with offline AOD?

- It would be nice to have, but do you have some concrete use-cases in mind? In general, to get 100% identical scores at AOD and MiniAOD level could take some additional work, and if there are no strong use-cases we might consider to postpone such development for post 11_2. 

- I'm thinking about DeepTau v1. Maybe it is good occasion to drop its support and simplify&clean-up the code? What do you think?

mbluj commented:
> > * Probably it can be useful to add "prediscriminat" as in other tauIDs to not run deepTau on useless taus.
> 
>     * I think for online pre-discriminant can depend on a path, so it can be added later after the new paths definitions are established. On the other hand, for offline AOD it makes sense to apply the same preselection as for miniAOD.

Yes, exactly. And it is why in "classic" tau discrimininants a prediscriminiant is a parameter where one can put one or a set of discriminants to be checked. One should check if the setup from the classic tau discriminants is what is fine here and if so copy it here.

> 
> > * Maybe it makes sense to allow also usage with offline AOD?
> 
>     * It would be nice to have, but do you have some concrete use-cases in mind? In general, to get 100% identical scores at AOD and MiniAOD level could take some additional work, and if there are no strong use-cases we might consider to postpone such development for post 11_2.
> 

Currently the only use case I have in mind is for testing. Then, deepTau could be run as part of tau sequence at AOD level, but we anyway think on moving it to MiniAOD so maybe it is not so useful.

>     * I'm thinking about DeepTau v1. Maybe it is good occasion to drop its support and simplify&clean-up the code? What do you think?

I think it is a good idea (same about DPF), but it can be simpler for reco guys to decouple the two things, i.e. firstly remove old deepTauIDs and then port to AOD. Do you agree?

kandrosov commented:
> I think it is a good idea (same about DPF), but it can be simpler for reco guys to decouple the two things, i.e. firstly remove old deepTauIDs and then port to AOD. Do you agree?

Yes, it makes sense.

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,357 @@
+
+    PFTauQualityCuts.primaryVertexSrc = cms.InputTag("hltPixelVertices")
+
+    process.chargedIsoPtSum = pfRecoTauDiscriminationByIsolation.clone(
```
I think it would be possible to use the new container format here to produce and store these isolation quantities in one go. Did you consider this? I did not check in detail, but probably you would only need two instances for the two dR versions. (see for example https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L68-L168)

lwezenbe replied:
I was not aware this, but I think it is a good idea. I have been trying to change this today but I am struggling a bit with the containers. To get the value of for example ChargedIsoPtSum, I would need the appropriate rawValue of the SingleTauDiscriminatorContainer, which is stored in a ValueMap. So in order to access the correct SingleTauDiscriminatorContainer, it seems like I need a Tau Ref. I was just wondering if/how this works when the TauCollection used is a edm::View of the reco::BaseTau.

lwezenbe replied:
It seems I was a bit too fast, I made a typo somewhere, but now I got the new container format working

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,357 @@
+        particleFlowSrc = cms.InputTag("hltParticleFlowReg"),
+        vertexSrc = PFTauQualityCuts.primaryVertexSrc,
+        Prediscriminants = cms.PSet(  BooleanOperator = cms.string( "and" ) ),
+        ApplyDiscriminationByECALIsolation = cms.bool(False),
```
In general, please don't redefine type of config arguments for basic types like cms.bool, cms.double... For options that are already set in the cloned object it is enough and preferred by RECO to just set the value (`... = False,`).

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,357 @@
+
+
+    #Add all sums and corrections for deepTauProducer
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.hpsPFTauPrimaryVertexProducer)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.hpsPFTauSecondaryVertexProducer)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.hpsPFTauTransverseImpactParameters)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.chargedIsoPtSum)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.chargedIsoPtSumdR03)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.neutralIsoPtSum)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.neutralIsoPtSumdR03)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.puCorrPtSum)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.footprintCorrection)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.footprintCorrectiondR03)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.neutralIsoPtSumWeight)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.neutralIsoPtSumWeightdR03)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.photonPtSumOutsideSignalCone)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.photonPtSumOutsideSignalConedR03)
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.hltFixedGridRhoFastjetAll)
```
I think it would be nicer to do this in a loop (using getattr such that you only need to write a list of the module names). However, if you used the container format resulting in less lines here, it might not be worth it. So don't take this as a strong request.

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,357 @@
+
+    wp_names = ["0.", "0."]
+
+    process.deepTauProducer = cms.EDProducer("DeepTauId",
+    taus                            = cms.InputTag("hltHpsPFTauProducerReg"),
+    pfcands                         = cms.InputTag('hltParticleFlowReg'),
+    vertices                        = cms.InputTag('hltPixelVertices'),
+    rho                             = cms.InputTag('hltFixedGridRhoFastjetAll'),
+    graph_file                      = cms.vstring(file_names),
+    mem_mapped                      = cms.bool(False),
+    version                         = cms.uint32(2),
+    debug_level                     = cms.int32(0),
+    disable_dxy_pca                 = cms.bool(True),
+    is_online                 		  = cms.bool(True),
+    pfTauTransverseImpactParameters = cms.InputTag('hpsPFTauTransverseImpactParameters'),
+    chargedIsoPtSum                 = cms.InputTag('chargedIsoPtSum'),
+    chargedIsoPtSumdR03             = cms.InputTag('chargedIsoPtSumdR03'),
+    neutralIsoPtSum                 = cms.InputTag('neutralIsoPtSum'),
+    neutralIsoPtSumdR03             = cms.InputTag('neutralIsoPtSumdR03'),
+    puCorrPtSum                     = cms.InputTag('puCorrPtSum'),
+    footprintCorrection             = cms.InputTag('footprintCorrection'),
+    neutralIsoPtSumWeight           = cms.InputTag('neutralIsoPtSumWeight'),
+    neutralIsoPtSumWeightdR03       = cms.InputTag('neutralIsoPtSumWeightdR03'),
+    photonPtSumOutsideSignalCone    = cms.InputTag('photonPtSumOutsideSignalCone'),
+    photonPtSumOutsideSignalConedR03 = cms.InputTag('photonPtSumOutsideSignalConedR03'),
```
Could you please run `scram build code-checks` and `scram build code-format`? I fear the latter will kill this almost perfect alignment here, but then we don't have to talk about fixing it. The checks and format applied by these scripts are anyway a requirement to pass the integration tests.

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,357 @@
+    #Add DeepTauProducer
+    process.HLTHPSMediumChargedIsoPFTauSequenceReg.insert(-1, process.deepTauProducer)
+ 
+    process.maxEvents = cms.untracked.PSet(
+        input = cms.untracked.int32(200)
+    )
```
Is this needed?

swozniewski commented on RecoTauTag/RecoTau/interface/DeepTauBase.h:
```diff
@@ -57,9 +62,10 @@ namespace deep_tau {
 
   class DeepTauBase : public edm::stream::EDProducer<edm::GlobalCache<DeepTauCache>> {
   public:
-    using TauType = pat::Tau;
+    using TauType = edm::View<reco::BaseTau>;
     using TauDiscriminator = reco::TauDiscriminatorContainer;
-    using TauCollection = std::vector<TauType>;
+    using TauCollection = TauType;
```
I think you can omit TauType and directly write `using TauCollection = edm::View<reco::BaseTau>;`

swozniewski commented on RecoTauTag/RecoTau/src/DeepTauBase.cc:
```diff
@@ -77,10 +82,13 @@ namespace deep_tau {
+        is_online(cfg.getParameter<bool>("is_online")),
         outputs_(outputCollection),
         cache_(cache) {
+
+
```
please remove new empty lines

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -982,14 +1270,28 @@ class DeepTauId : public deep_tau::DeepTauBase {
     event.getByToken(rho_token_, rho);
 
     tensorflow::Tensor predictions(tensorflow::DT_FLOAT, {static_cast<int>(taus->size()), deep_tau::NumberOfOutputs});
+
+
     for (size_t tau_index = 0; tau_index < taus->size(); ++tau_index) {
+
```
please don't add extra empty lines here

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -982,14 +1270,28 @@ class DeepTauId : public deep_tau::DeepTauBase {
-        getPredictionsV2(taus->at(tau_index), *electrons, *muons, *pfCands, vertices->at(0), *rho, pred_vector);
-      else
-        throw cms::Exception("DeepTauId") << "version " << version << " is not supported.";
+      if (is_online){
+        if (version == 1)
+          getPredictionsV1<reco::PFCandidate, reco::PFTau>(taus->at(tau_index), tau_index, electrons, muons, pred_vector, tauIDs);
+        else if (version == 2){
+          getPredictionsV2<reco::PFCandidate, reco::PFTau>(taus->at(tau_index), tau_index, electrons, muons, *pfCands, vertices->at(0), *rho, pred_vector, tauIDs);
+        }
+        else
+          throw cms::Exception("DeepTauId") << "version " << version << " is not supported.";
+      } else {
+        if (version == 1)
+          getPredictionsV1<pat::PackedCandidate, pat::Tau>(taus->at(tau_index), tau_index, electrons, muons, pred_vector, tauIDs);
+        else if (version == 2)
+          getPredictionsV2<pat::PackedCandidate, pat::Tau>(taus->at(tau_index), tau_index, electrons, muons, *pfCands, vertices->at(0), *rho, pred_vector, tauIDs);
+        else
+          throw cms::Exception("DeepTauId") << "version " << version << " is not supported.";
+      }
```
Maybe it would be better to put the if-statements the other way round in order to avoid the duplication of the exception and to have versions (also future ones) in a single place each.

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1069,6 +1378,7 @@ class DeepTauId : public deep_tau::DeepTauBase {
       if (dR2 < outer_dR2)
         addObject(n, deta, dphi, outer_grid);
     }
+
```
please remove

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1092,19 +1402,22 @@ class DeepTauId : public deep_tau::DeepTauBase {
                       {"outer_all_dropout_4/Identity"},
                       &pred_vector);
     }
+
```
please remove

swozniewski commented:
Dear @lwezenbe , I viewed most parts but I'm not yet completely through DeepTauId.cc and will need to take a further look next week.

One major change that I would request (as mentioned in one of my previous comments) is to use the new tauID format and the corresponding producer for the isolation quantities. We should not introduce new use cases for the old format. Actually, we are requested to remove the ones that still exist at some point.
This would also save many lines of code in this development. The only thing that is a bit more tricky is to find the right indices. This requires to read the config of the tau ID input object via provenance. You can find an example here:
https://github.com/cms-sw/cmssw/blob/ffbf7f8ea21b840022b4744aa36eb89ca3c4fe1a/DQMOffline/Trigger/plugins/HLTTauRefProducer.cc#L165-L173
Actually, since you are setting up this config of the ID producer only for deepTau, I think it would be also fine to hand over the indices via the python config file, ideally in a way that nobody changes them differently for the input ID producer and the deepTauID producer by accident. What do you think @mbluj ? A compromise could also be to hardcode the indices in the deepTauID Producer and only check via provenance whether they are correct (in case sb would change the config in the future). Without the logic to search for the indices, this would be a bit shorter.

Other useful links concerning this topic:
The definition of the format is here: https://github.com/cms-sw/cmssw/blob/49acac1cee9f9d86c4f7223614d21b547127cd38/DataFormats/TauReco/interface/TauDiscriminatorContainer.h
Example for the configuration of the ID producer:  https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L68-L168 (same link as in my first comment)

Please don't hesitate to ask if sth is not clear about this.

mbluj commented:
Hello,

firstly I would like to admit that I have not time to revise this development, yet. I see that a great review have been already preformed by @swozniewski - thanks! I agree with Sebastian's comments.
> 
> One major change that I would request (as mentioned in one of my previous comments) is to use the new tauID format and the corresponding producer for the isolation quantities. We should not introduce new use cases for the old format. Actually, we are requested to remove the ones that still exist at some point.
> This would also save many lines of code in this development. The only thing that is a bit more tricky is to find the right indices. This requires to read the config of the tau ID input object via provenance. You can find an example here:
> https://github.com/cms-sw/cmssw/blob/ffbf7f8ea21b840022b4744aa36eb89ca3c4fe1a/DQMOffline/Trigger/plugins/HLTTauRefProducer.cc#L165-L173
> Actually, since you are setting up this config of the ID producer only for deepTau, I think it would be also fine to hand over the indices via the python config file, ideally in a way that nobody changes them differently for the input ID producer and the deepTauID producer by accident. What do you think @mbluj ? A compromise could also be to hardcode the indices in the deepTauID Producer and only check via provenance whether they are correct (in case sb would change the config in the future). Without the logic to search for the indices, this would be a bit shorter.
> 
> Other useful links concerning this topic:
> The definition of the format is here: https://github.com/cms-sw/cmssw/blob/49acac1cee9f9d86c4f7223614d21b547127cd38/DataFormats/TauReco/interface/TauDiscriminatorContainer.h
> Example for the configuration of the ID producer: https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/HPSPFTaus_cff.py#L68-L168 (same link as in my first comment)
> 

I agree that use of the new dataformat will be useful also in the context of HLT. It will allow to save some computations (all pt-sums will be computed with the same loop), also interface will be simpler. I have not idea about usage of provenance at HLT, but I do not see any reason that it will not work. If so, I propose to use it to configure deepTau for HLT as patTaus are configured. This can however require that HLT paths are setup in a bit different way than they are now being setup, but it should not be a problem. @swozniewski - it is possible to use pf-tau selectors with the new dataformat?

In principle the new dataformat can be also used in output - deepTau (offline version for miniAOD) produces 3 raw (float) values (DNN scores) and a big set of booleans (WPs).

swozniewski commented:
yes, it is possible to use PFTauSelector. It accepts the old and the new format. Do you mean for HLT in general? Or is there a use case for this development?

The output of deepTau is already in the new format.

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,357 @@
+    				'outer:RecoTauTag/TrainingFiles/data/DeepTauId/deepTau_2017v2p6_e6_outer.pb',
+    			]
+
+    wp_names = ["0.", "0."]
```
btw. talking about the output, this is a remainder from testing, right?

swozniewski replied:
Is this to be kept like this for now?

mbluj commented:
> yes, it is possible to use PFTauSelector. Do you mean for HLT in general? Or is there a use case for this development?

OK, thanks for confirmation.
I mean that usage of isolation pt-sums stored in the new dataformat for deepTau for HLT will probably trigger migration of all tau HLT paths to the new dataformat regardless if all these paths will use of deepTau. The reason is two fold: shared usage of modules is preferred at HLT as it saves resources, and because it is simpler to maintain. For the migration PFTauSelector must be compatible with the new dataformat.

> The output of deepTau is already in the new format.

OK, great.

swozniewski commented:
True. Right, I prepared the PFTauSelector for the new format when I realized that this would be needed for HLT and still thought that the update of HLT would be doable in the same development...

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1283,6 +1594,7 @@ class DeepTauId : public deep_tau::DeepTauBase {
     const bool valid_index_pf_gamma = cell_map.count(CellObjectType::PfCand_gamma);
     const bool valid_index_ele = cell_map.count(CellObjectType::Electron);
 
+
```
please remove

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -2078,9 +2396,22 @@ class DeepTauId : public deep_tau::DeepTauBase {
+  edm::EDGetTokenT<reco::PFTauDiscriminator> photonPtSumOutsideSignalConedR03_inputToken;
+  edm::EDGetTokenT<edm::AssociationVector<reco::PFTauRefProd, std::vector<reco::PFTauTransverseImpactParameterRef>>> PFTauTransverseImpactParameters_token;
+
+
```
please remove

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -1001,30 +1303,37 @@ class DeepTauId : public deep_tau::DeepTauBase {
-    const tensorflow::Tensor& inputs = createInputsV1<dnn_inputs_2017v1>(tau, electrons, muons);
+  template <typename CandidateCastType, typename TauCastType>
+  void getPredictionsV1(const TauType::const_reference& tau,
+                        size_t tau_index,
```
Here, and in all similar methods: Is there a reason to have tau_index as non-const argument? At least I did not notice any. If not, please turn into const.

lwezenbe replied:
There is not, this has been changed.

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -414,6 +414,218 @@ namespace {
+    const reco::PFTauDiscriminator& photonPtSumOutsideSignalConedR03Collection;
+    edm::Handle<edm::AssociationVector<reco::PFTauRefProd, std::vector<reco::PFTauTransverseImpactParameterRef>>> PFTauTransverseImpactParameters;
+
+    const float getChargedIsoPtSum(const reco::PFTau& tau, size_t tau_index) {return chargedIsoPtSumCollection.value(tau_index);}
+    const float getChargedIsoPtSum(const pat::Tau& tau, size_t tau_index)  {return tau.tauID("chargedIsoPtSum");}
+
+    const float getChargedIsoPtSumdR03(const reco::PFTau& tau, size_t tau_index) {  return chargedIsoPtSumdR03Collection.value(tau_index);}
+    const float getChargedIsoPtSumdR03(const pat::Tau& tau, size_t tau_index) {  return tau.tauID("chargedIsoPtSumdR03");}
+
+    const float getFootprintCorrectiondR03(const reco::PFTau& tau, size_t tau_index) {  return footprintCorrectionCollection.value(tau_index);}
+    const float getFootprintCorrectiondR03(const pat::Tau& tau, size_t tau_index) {     return tau.tauID("footprintCorrectiondR03");}
+
+
+    const float getNeutralIsoPtSum(const reco::PFTau& tau, size_t tau_index) {  return neutralIsoPtSumCollection.value(tau_index);}
+    const float getNeutralIsoPtSum(const pat::Tau& tau, size_t tau_index) {  return tau.tauID("neutralIsoPtSum");}
+
+    const float getNeutralIsoPtSumdR03(const reco::PFTau& tau, size_t tau_index) {  return neutralIsoPtSumdR03Collection.value(tau_index);}
+    const float getNeutralIsoPtSumdR03(const pat::Tau& tau, size_t tau_index)   {return tau.tauID("neutralIsoPtSumdR03");}
+
+
+    const float getNeutralIsoPtSumWeight(const reco::PFTau& tau, size_t tau_index) { return neutralIsoPtSumWeightCollection.value(tau_index);}
+    const float getNeutralIsoPtSumWeight(const pat::Tau& tau, size_t tau_index) {return tau.tauID("neutralIsoPtSumWeight");}
+
+
+    const float getNeutralIsoPtSumdR03Weight(const reco::PFTau& tau, size_t tau_index) { return neutralIsoPtSumWeightdR03Collection.value(tau_index);}
+    const float getNeutralIsoPtSumdR03Weight(const pat::Tau& tau, size_t tau_index) {return tau.tauID("neutralIsoPtSumWeightdR03");}
+
+
+    const float getPhotonPtSumOutsideSignalCone(const reco::PFTau& tau, size_t tau_index) { return photonPtSumOutsideSignalConeCollection.value(tau_index);}
+    const float getPhotonPtSumOutsideSignalCone(const pat::Tau& tau, size_t tau_index) {return tau.tauID("photonPtSumOutsideSignalCone");}
+
+
+    const float getPhotonPtSumOutsideSignalConedR03(const reco::PFTau& tau, size_t tau_index) {  return photonPtSumOutsideSignalConedR03Collection.value(tau_index);}
+    const float getPhotonPtSumOutsideSignalConedR03(const pat::Tau& tau, size_t tau_index) {return tau.tauID("photonPtSumOutsideSignalConedR03");}
+
+
```
Is there a pattern behind the number of empty lines? Please explain or make it more consistent.

lwezenbe replied:
This was initially to make it more clear to myself what functions were already available but I forgot to remove the empty lines afterwards. This has been done now.

swozniewski commented:
Ok, I think these are all comments from my side for now. Thanks a lot for the nice work so far!

lwezenbe commented:
I forgot to mention in the last commit message that for the change to the new container format, I checked if the values for the different values of chargedIsoPtSum and so on changed as well. The footPrintCorrection changed values (previously apparently having value 0) which was tracked down to the footprintcorrections in the pfRecoTauDiscriminationByIsolation_cfi.py file being an empty parameter set in the previous setup. There were also some changed values in NeutralIsoPtSumWeight, the origin of which was not yet found. 

On a side note, I was also implementing the prediscriminator but ran into an issue when defining the TauDiscInfo as done here (https://github.com/cms-tau-pog/cmssw/blob/e189ecad47a8d1358c531e15610686f7429d784d/RecoTauTag/RecoTau/interface/TauDiscriminationProducerBase.h#L82) so it is not yet implemented in this version. I will return to this issue tomorrow and try to push a version including prediscriminants.

swozniewski commented:
Hi Liam, I'm not sure whether I understand you correctly, so maybe I'm missing your point with this answer: The base configs (_cfi) are not necessarily completely configured. What is prodiced for the pat taus in the end, is defined in the HPSPFTaus_cff.py file (before and after the introduction of the new format). So this is probably the best reference to configure your rerun ID inputs for deepTau. These IDs did not change with the switch of the formats (apart from a small difference in the combinedIso, but I think you don't use this), so I'm a bit surprised about the differences you are reporting. Is it because you were using the configs from the cfi files?

lwezenbe commented:
Hello, I realize my response was a bit vague. I was referring to running the module on RECO. The update  function in the deepTauAtHLT.py config file has been changed so it provides the input for the discriminator container as in HPSPFTaus_cff.py. I removed the wp list in the config and so on because the input for the deeptau just uses the raw values. Before I was using modules as defined in the code by Maria Teresa (https://github.com/mgrippo/cmssw/blob/deepTau_hlt_dev/HLTrigger/Configuration/python/HLT_User_cff.py#L13774) and now I have made the switch to the input like in HPSPFTaus_cff.py but one of the differences is that the different modules separately had some different inputs for different quantities. For example, the footprintcorrection specified a prediscriminant based on decay mode while most of the others modules did not specify a prediscriminant.
What I meant with my comment on tests is that since I was not expecting major differences, I checked what the raw values were with both ways of getting input. The differences are minor and only when running on RECO and mostly understood what parameter change caused them. I just wanted to be transparent that if anyone was running the code on their side, a small change might be observed in the output file now compared to the previous version.

swozniewski commented:
If I understand correctly, the inputs configured by Maria Teresia for HLT are different than the ones entering usual deepTau at miniAOD. But these were never supposed to enter deepTau, right? My feeling is that we should use inputs matching the training conditions, which is I guess the case now. You may have already discussed it internally, probably with @kandrosov . So just in case, please make sure that the choice of the configuration is well motivated.

kandrosov commented:
The definition of modules in Maria Teresa's code is also supposed to be a copy from HPSPFTaus_cff with some minimal modifications to make it run at HLT. So, in case of any differences, use HPSPFTaus_cff version as a reference (if it works), because it should have conditions that are closer to the ones used for the training.

lwezenbe commented:
Prediscriminants have now been added. Unfortunately due to other obligations my work on this had slowed down quite a bit, but I will try to put my focus on this project to hopefully wrap up the changes this week. I have a few questions however:

> Generally PFTaus can be also build with pat::PackedCandidates, so maybe one should also consider it in this code

Do you have an example of how this done?

> Maybe it makes sense to allow also usage with offline AOD?

Should I still try to implement this or should we postpone this change for post 11_2_X changes?

> Actually, since you are setting up this config of the ID producer only for deepTau, I think it would be also fine to hand over the indices via the python config file, ideally in a way that nobody changes them differently for the input ID producer and the deepTauID producer by accident. What do you think @mbluj ? A compromise could also be to hardcode the indices in the deepTauID Producer and only check via provenance whether they are correct (in case sb would change the config in the future). Without the logic to search for the indices, this would be a bit shorter.

> I agree that use of the new dataformat will be useful also in the context of HLT. It will allow to save some computations (all pt-sums will be computed with the same loop), also interface will be simpler. I have not idea about usage of provenance at HLT, but I do not see any reason that it will not work. If so, I propose to use it to configure deepTau for HLT as patTaus are configured. This can however require that HLT paths are setup in a bit different way than they are now being setup, but it should not be a problem. @swozniewski - it is possible to use pf-tau selectors with the new dataformat?

Currently the indices for the discriminators are give in the config file, but not really in a way that nobody can change them differently without realizing. I am looking into using provenance to get the correct indices and hopefully will have a first version for that pushed by the end of the day. However, I am not sure I understood the rest of the conversation that stemmed from these quotes. Is the PFTauSelector something I should look into for the deepTau at HLT or was this a side note? If the first is the case, could you elaborate a bit more on this?

Lastly, on the input configuration, this is still in a testing mode, I will make sure to check that everything is well modeled after HPSPFTaus_cff and I will soon change other things like the definition of the working points in the config to be the correct input.

mbluj commented:
> I have a few questions however:
> 
> > Generally PFTaus can be also build with pat::PackedCandidates, so maybe one should also consider it in this code
> 
> Do you have an example of how this done?

The reco::PFTaus can be produced not only on top of AOD inputs, i.e. reco::PFCandidates, but also on top of miniAOD inputs, i.e. pat::PackedCandidates, and hence the two types of candidates can be returned as PFTau signal or isolation constituents.. The latter production is currently not used, but there is plan to enable it. The two type of candidates have a bit different interfaces which is taken into account in several places in tau code and probably can be also accounted for in deepTau. For example in method getEmFraction(reco::PFTau) the lead candidate is dynamically cast to be reco::PFCadindate, but in case it is not it is not checked if it is packedCandidate (should be) and thus a proper handling of this case is missing. (Similar in other methods). 
In addition it can be worth to check if reco::PFCadindate->bestTrack() returns pointer to the same track as packedCandidate->bestTrack() (relevant for muons and electrons which can have more than one associated track).

> > Maybe it makes sense to allow also usage with offline AOD?
> 
> Should I still try to implement this or should we postpone this change for post 11_2_X changes?

It depends on effort to do so. There are pros and cons of gradual changes - what do you think? 

> 
> > Actually, since you are setting up this config of the ID producer only for deepTau, I think it would be also fine to hand over the indices via the python config file, ideally in a way that nobody changes them differently for the input ID producer and the deepTauID producer by accident. What do you think @mbluj ? A compromise could also be to hardcode the indices in the deepTauID Producer and only check via provenance whether they are correct (in case sb would change the config in the future). Without the logic to search for the indices, this would be a bit shorter.
> 
> > I agree that use of the new dataformat will be useful also in the context of HLT. It will allow to save some computations (all pt-sums will be computed with the same loop), also interface will be simpler. I have not idea about usage of provenance at HLT, but I do not see any reason that it will not work. If so, I propose to use it to configure deepTau for HLT as patTaus are configured. This can however require that HLT paths are setup in a bit different way than they are now being setup, but it should not be a problem. @swozniewski - it is possible to use pf-tau selectors with the new dataformat?
> 
> Currently the indices for the discriminators are give in the config file, but not really in a way that nobody can change them differently without realizing. I am looking into using provenance to get the correct indices and hopefully will have a first version for that pushed by the end of the day. However, I am not sure I understood the rest of the conversation that stemmed from these quotes. Is the PFTauSelector something I should look into for the deepTau at HLT or was this a side note? If the first is the case, could you elaborate a bit more on this?

I wanted to check with Sebastian if other modules used at HLT (PFTauSelectors) are compatible with the new data format used by tau discriminators - fortunately it is the case (otherwise it will require another development beyond this PR).
 
I hope that my comments are more clear now.

lwezenbe commented:
> The reco::PFTaus can be produced not only on top of AOD inputs, i.e. reco::PFCandidates, but also on top of miniAOD inputs, i.e. pat::PackedCandidates, and hence the two types of candidates can be returned as PFTau signal or isolation constituents.. The latter production is currently not used, but there is plan to enable it. The two type of candidates have a bit different interfaces which is taken into account in several places in tau code and probably can be also accounted for in deepTau. For example in method getEmFraction(reco::PFTau) the lead candidate is dynamically cast to be reco::PFCadindate, but in case it is not it is not checked if it is packedCandidate (should be) and thus a proper handling of this case is missing. (Similar in other methods).
In addition it can be worth to check if reco::PFCadindate->bestTrack() returns pointer to the same track as packedCandidate->bestTrack() (relevant for muons and electrons which can have more than one associated track).

I see, so you're saying that whenever a candidate is used for something, it should take into account that the input taus could also be built from pat::PackedCandidates? I will try to change this now.

> It depends on effort to do so. There are pros and cons of gradual changes - what do you think? 

I think it would be nice but that we might be a bit short on time to do it now. It was just repeated in the TSG meeting that there are now only three weeks to implement all the changes for the pre8 version (which is the last chance to get something into cmssw_11_2). So I believe it would be good wrap this up rather sooner than later after everything that's needed for the first version is implemented and tested and to return to this later.

Apart from the candidates, are there other changes or tests that should be performed now? (I will implement the candidate part now and change the deepTauAtHLT.py config file to not contain lines from testing anymore)

mbluj commented:
Due to time constraints I propose to postpone both candidate part and full port to offline RECO/AOD inputs (including electrons and muons) for further development (the candidate part is also irrelevant for HLT). You can add "TODO" comments to the code.
What concerns tests I do not request any additional ones. @swozniewski - are you fine with current status of the development?
N.B. We should keep in mind that Christian also wants implement some changes on top of these (with a different PR).

swozniewski commented:
I'll have a look. @lwezenbe maybe you said it already and I missed it: Apart from what Michal suggests to postpone, are all envisaged updates finished?

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,187 @@
+    process.hpsPFTauBasicDiscriminators = pfRecoTauDiscriminationByIsolation.clone(
+        PFTauProducer = cms.InputTag('hltHpsPFTauProducerReg'),
+        Prediscriminants = requireDecayMode.clone(),
+        # Prediscriminants = cms.PSet(  BooleanOperator = cms.string( "and" ) ),
```
Is there a use case for reactivating this line? If not or unlikely, please remove.

swozniewski commented on HLTrigger/Configuration/python/deepTauAtHLT.py:
```diff
@@ -0,0 +1,187 @@
+
+    ## Cut based isolations dR=0.3
+    process.hpsPFTauBasicDiscriminatorsdR03 = process.hpsPFTauBasicDiscriminators.clone(
+        customOuterCone = cms.double(0.3)
```
-> `...= 0.3`

swozniewski commented on RecoTauTag/RecoTau/plugins/DeepTauId.cc:
```diff
@@ -414,6 +414,374 @@ namespace {
     }
   }  // namespace dnn_inputs_2017_v2
 
+  const std::map<std::string, size_t> matchDiscriminatorIndices(
```
I think it would be good to avoid calling this map with strings on an event base. Maybe this could be translated to an enum within this function. This way it could already be checked whether all needed inputs are present, as well.

lwezenbe replied:
I agree that it would be good to avoid this. I am not sure I am following 100% what you mean with the enum solution however. Could you elaborate on this?

swozniewski replied:
True, I did not fully express my thought, also because I'm not sure that it is the best way to do it and I did not fully develop it. Thinking about it again, how about the following?: You can define an enum within the DeepTauID class listing the cutbased ID types. In the end of the matchDiscriminatorIndices function, you could call for each enum entry your map with the corresponding string (hardcoded, changing those would anyway require code changes in this class) and fill another map<enum,index> and return this one instead. This way you could already check, whether all needed inputs are available.
Alternatively and a bit shorter you could add some if-else to the matchDiscriminatorIndices to translate the strings from provenance into enum in order to fill the map. This would be a bit more direct, but also without check of completeness. But I'm also not sure how useful such a check would be.

swozniewski replied:
btw. with an enum you don't even need a map. You can use a vector of given size right away. This could be even more efficient. Just to be clear: The purpose of the enum would be to have human readible calls in your helper functions where you load the ID inputs.
If there was only one dR version, it might be even easier to define one index variable as class member for each ID type. But having this for each dR version would become a bit messy, I guess.
But you see that there are quite some options. Maybe you find something more beatiful.

swozniewski replied:
With 'check' I don't necessarily mean some super nice error message. The behaviour so far, where you would get an error that the map does not contain a certain string is actually fine already (I thought it's maybe nicer to get that during the initialization already, but in the event loop is actually fine as well). That's maybe the advantage of the map. If you switch to a vector, you'd need to take some extra care that all entries are filled properly in order to avoid undefined behavior.

swozniewski replied:
I hope I didn't increase confusion with my loose thoughts now.

lwezenbe replied:
No, no problem, the input is definitely welcome. I was just thinking on how to properly do this. I am currently just struggling a bit to find a way to do this in a nice manner. Personally I think that the explicit string call in the maps in the helper functions make it fairly readable, but I do see your point.

swozniewski replied:
With the enum, your helper functions would look almost the same, just without the quotes around the names.

swozniewski replied:
that's why I suggested enum. For readability reasons, I would not like calls via plain intergers or sth like that, either.

lwezenbe replied:
Ah yes, I just realized I was completely misinterpreting something in your explanation. I think we are on the same page now and I will make the changes

swozniewski commented:
Ok I went through it quickly and only have a small number of minor comments (see above). I think you solved it very nicely with the now ID format.

lwezenbe commented:
> Apart from what Michal suggests to postpone, are all envisaged updates finished?

The main updates that I'm aware of were adding prediscriminants, implementing the new container format and getting the indices for it in a way that's more robust with regards to changes in the config file. All of these are implemented now. If I forgot to mention something here or just forgot about something in general, please let me know. 

> Ok I went through it quickly and only have a small number of minor comments (see above)

I will go through them this morning and push these changes (hopefully) this morning. I just have one last thing that I am not completely sure about regarding the working points. @kandrosov Right now I used some simple simple_value purely for testing purposes. How should the working points be defined in the config file? Should this be done the same way as it is done offline as here?
https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L660-L686

kandrosov commented:
> Right now I used some simple simple_value purely for testing purposes. How should the working points be defined in the config file? Should this be done the same way as it is done offline as here?
> https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/python/tools/runTauIdMVA.py#L660-L686

To define working points, the dedicated studies should be done and the values can vary for different paths. Such studies will happen after the code is integrated into CMSSW, so for the moment use some placeholder values.

swozniewski commented:
Ok, thanks for confirmation. Then no further action on the concerned line needed.

lwezenbe commented:
I added the translation of the discriminator index map to enum as requested as well as a check if all required discriminators are there. I have also added a check which makes sure the indices are not assigned for every event but only on the first.
The push itself took a little longer than expected because I noticed that in the online output file, I found no extra collection for deeptau variables while there were output collections from all the other producers defined in deepTauAtHLT.py. After looking into this more, the reason is still not clear to me.
I hope to have this resolved on monday, I will then also rerun the battery of tests requested on the "workflow for pull request" page. I had already run them after my initial push and there were no issues then.

swozniewski commented:
Do you in some way include this config in your test run?
https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/RecoTauTag_EventContent_cff.py
If yes, you may need to add deeptau in one of the lists.

mbluj commented:
> Do you in some way include this config in your test run?
> https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/RecoTauTag_EventContent_cff.py
> If yes, you may need to add deeptau in one of the lists.

Yes, to produce some product with current "unscheduled" framework one has to setup a consumer of the product, i.e. other module (can be an output module) using it. 
BTW, please share your test configuration, it will be useful. Maybe it can be even added to `RecoTauTag/RecoTau/test` directory - @swozniewski what do you think?

swozniewski commented:
> BTW, please share your test configuration, it will be useful. Maybe it can be even added to `RecoTauTag/RecoTau/test` directory - @swozniewski what do you think?

yes, good point

swozniewski commented on RecoTauTag/RecoTau/interface/DeepTauBase.h:
```diff
@@ -115,6 +115,28 @@ namespace deep_tau {
+      PUcorrPtSum
+    };
+
+    std::map<BasicDiscriminator, std::string> stringFromDiscriminator_{
+        {ChargedIsoPtSum, "ChargedIsoPtSum"},
+        {NeutralIsoPtSum, "NeutralIsoPtSum"},
+        {NeutralIsoPtSumWeight, "NeutralIsoPtSumWeight"},
+        {FootprintCorrection, "TauFootprintCorrection"},
+        {PhotonPtSumOutsideSignalCone, "PhotonPtSumOutsideSignalCone"},
+        {PUcorrPtSum, "PUcorrPtSum"}};
+
+    std::vector<BasicDiscriminator> requiredBasicDiscriminators = {
+        ChargedIsoPtSum, NeutralIsoPtSum, NeutralIsoPtSumWeight, PhotonPtSumOutsideSignalCone, PUcorrPtSum};
+    std::vector<BasicDiscriminator> requiredBasicDiscriminatorsdR03 = {
+        ChargedIsoPtSum, NeutralIsoPtSum, NeutralIsoPtSumWeight, PhotonPtSumOutsideSignalCone, FootprintCorrection};
```
please make these `static const`

swozniewski commented on RecoTauTag/RecoTau/interface/DeepTauBase.h:
```diff
@@ -89,17 +101,45 @@ namespace deep_tau {
+    edm::EDGetTokenT<CandidateType> pfcandToken_;
     edm::EDGetTokenT<reco::VertexCollection> vtxToken_;
     std::map<std::string, WPList> workingPoints_;
+    const bool is_online;
     OutputCollection outputs_;
     const DeepTauCache* cache_;
+
+    static const std::map<BasicDiscriminator, std::string> stringFromDiscriminator_;
+    static const std::vector<BasicDiscriminator> requiredBasicDiscriminators;
+    static const std::vector<BasicDiscriminator> requiredBasicDiscriminatorsdR03;
```
I just realized that not all members have an underscore at the end. Could you please add that in order to stick to the naming conventions? (is_online and the two vectors)

swozniewski replied:
... and good point to move that part to the private sector!

lwezenbe commented:
I believe I have now made all the requested changes.

> Do you in some way include this config in your test run?
https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/Configuration/python/RecoTauTag_EventContent_cff.py
If yes, you may need to add deeptau in one of the lists.

This was indeed what I was forgetting. After doing this, the output was according to expectations.

> BTW, please share your test configuration, it will be useful. Maybe it can be even added to RecoTauTag/RecoTau/test directory

Concerning the tests. To perform the tests on AOD, I ran the already existing runDeepTauIDsOnMiniAOD.py:
https://github.com/cms-sw/cmssw/blob/master/RecoTauTag/RecoTau/test/runDeepTauIDsOnMiniAOD.py
This uses the same module that I was adapting and the input in the configuration was written so it does not need to be adapted for AOD.
For test with RECO, I simply used hltConfiguration to get a defined fragment for a ditau path HLT_User_cff.py with which I then used cmsDriver to create a config file that uses the update function from deepTauAtHLT.py

I have also performed a check on one last thing that I had doubts about, namely the values of the basicDiscriminators changing at some point without changing the config file. I now traced this back to the difference between pre3 and pre5. Using the exact same setup (the latest version pushed here), I ran HLT in a pre3 and a pre5 environment, and the changes I observed occured.

Lastly, I have performed the tests on this page with the last setup (before the final name changes requested):
http://cms-sw.github.io/PRWorkflow.html
The basic battery of tests was fully passed in the pre3 environment, no issues there.
When performing this in pre5, I got some errors while running
`git cms-checkdeps -a -A`
but they seem to have nothing to do with the code I changed. It is simply complaining about some files being removed from the package:
".../CMSSW_11_2_0_pre5/poison/DataFormats/GeometryCommonDetAlgo/interface/LocalErrorExtended.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
 #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
"

swozniewski commented:
Thanks a lot for the follow up and the careful checks! I think we can move on then if you agree @mbluj 
I have no idea what could have caused changes in the basic IDs. @lwezenbe could you clarify which of the basic IDs change? All of them?

mbluj commented:
> 
> 
> Thanks a lot for the follow up and the careful checks! I think we can move on then if you agree @mbluj

Yes, I agree. However, I propose prepare a test config to run AOD version of deepTau with standard AOD and e.g. to store it within pat::Tau. It can be modified version of `PhysicsTools/PatAlgos/test/patTuple_onlyTaus_cfg.py`. It can be even added after a PR to official CMSSW is prepared, but I am sure that reco people will ask us for such test.

> I have no idea what could have caused changes in the basic IDs. @lwezenbe could you clarify which of the basic IDs change? All of them?

It is indeed strange.

lwezenbe commented:
> @lwezenbe could you clarify which of the basic IDs change? All of them?

Yes, as far as I can see, it is all of them. One thing I can add here for the full picture is that when I run the latest setup in a pre3 environment, I have to manually add the Basic2DClusterForEachSeed plugin to RecoParticleFlow.PFClusterProducer or I get a 'pluginNotFound' error, which does not happen in pre5 or pre6. Other than this, I started from a clean install of CMSSW for both environments and just cloned the latest setup in both of them.

> Yes, I agree. However, I propose prepare a test config to run AOD version of deepTau with standard AOD and e.g. to store it within pat::Tau. It can be modified version of PhysicsTools/PatAlgos/test/patTuple_onlyTaus_cfg.py. It can be even added after a PR to official CMSSW is prepared, but I am sure that reco people will ask us for such test.

Ok, I will try to do this soon, but I will probably not have the time for it in the next few days.

lwezenbe commented:
I have not had time to prepare a test config but I wanted to check with you what the next step should be. It's the first time I'm actually trying to merge something into the CMSSW framework so I am not sure where we go from here. Should I prepare a new pull request?

swozniewski commented:
As soon as you have added the test config, I will merge this PR into the tau-pog branch. If necessary I will rebase it to current master, rerun the integration tests and create a PR to official CMSSW from there. Then RECO conveners will review the PR. I will try to address smaller comments myself. If major changes are requested, I'll get back to you and we will see how to proceed. As soon as RECO conveners are satisfied, they merge the PR into CMSSW and we are hopefully done :) (sometimes issues are detected even after that).

mbluj commented:
@swozniewski: I propose to merge it now before the test config is here to prepare PR to the official CMSSW before this WE if possible. The config can be added at later stage by a direct commit to the tau-pog repository (or additional small PR to this branch), I think. I suggest doing this in that way as 112X will be closed soon and it will be good to have this PR in it. There is also a pending development by Christain on top of it. Do you agree?

swozniewski commented:
Ok, I agree. I'm short of time today, but I'll try. Otherwise I will take care of it tomorrow.

kandrosov commented:
I agree with Michal that if possible we should try to enter in 11_2.
@lwezenbe @mbluj @swozniewski thanks a lot for following up this PR during the last month and making the code production-ready!
